hbase-server/src/main/java/org/apache/hadoop/hbase/master/MobFileCleanerChore.java

```diff
@@ -0,0 +1,260 @@
+                sf.initReader();
+                byte[] mobRefData = sf.getMetadataValue(HStoreFile.MOB_FILE_REFS);
+                byte[] mobCellCountData = sf.getMetadataValue(HStoreFile.MOB_CELLS_COUNT);
+                byte[] bulkloadMarkerData = sf.getMetadataValue(HStoreFile.BULKLOAD_TASK_KEY);
+                if (mobRefData == null
+                    && (mobCellCountData != null || bulkloadMarkerData == null)) {
+                  LOG.info("Found old store file with no MOB_FILE_REFS: {} - "
+                    + "can not proceed until all old files will be MOB-compacted.", pp);
+                  return;
+                } else if (mobRefData == null) {
+                  LOG.info("Skipping file without MOB references (can be bulkloaded file):{}", pp);
+                  continue;
+                }
+                String[] mobs = new String(mobRefData).split(",");
+                regionMobs.addAll(Arrays.asList(mobs));
+              }
+            } catch (FileNotFoundException e) {
+              // TODO
+              LOG.warn(e.getMessage());
+              continue;
+            }
+            succeed = true;
+          }
+          // Add MOB refs for current region/family
+          allActiveMobFileName.addAll(regionMobs);
+        } // END column families
+      } // END regions
+      // Check if number of MOB files too big (over 1M)
+      if (allActiveMobFileName.size() > 1000000) {
+        LOG.warn("Found too many active MOB files: {}, this may result in high memory pressure.", 
+          allActiveMobFileName.size());
+      }
+      // Now scan MOB directories and find MOB files with no references to them
+      long now = System.currentTimeMillis();
+      long minAgeToArchive = conf.getLong(MobConstants.MOB_MINIMUM_FILE_AGE_TO_ARCHIVE_KEY,
+        MobConstants.DEFAULT_MOB_MINIMUM_FILE_AGE_TO_ARCHIVE);
+      for (ColumnFamilyDescriptor hcd : list) {
+        List<Path> toArchive = new ArrayList<Path>();
+        String family = hcd.getNameAsString();
+        Path dir = MobUtils.getMobFamilyPath(conf, table, family);
+        RemoteIterator<LocatedFileStatus> rit = fs.listLocatedStatus(dir);
+        while (rit.hasNext()) {
+          LocatedFileStatus lfs = rit.next();
+          Path p = lfs.getPath();
+          if (!allActiveMobFileName.contains(p.getName())) {
+            // MOB is not in a list of active references, but it can be too
+            // fresh, skip it in this case
+            /* DEBUG */ LOG.debug(
+              " Age=" + (now - fs.getFileStatus(p).getModificationTime()) + " MOB file=" + p);
+            if (now - fs.getFileStatus(p).getModificationTime() > minAgeToArchive) {
+              toArchive.add(p);
+            } else {
+              LOG.debug("Skipping fresh file: {}", p);
+            }
+          }
+        }
+        LOG.info(" MOB Cleaner found {} files for family={}", toArchive.size() , family);
```

Comment:
if we get to this point and it's taken us longer than `MOB_MINIMUM_FILE_AGE_TO_ARCHIVE_KEY` then we need to do something, since we know that there can easily be a disconnect between the view we had of regions while building the set of active mob files and current.

1) We could log an ERROR and abort. Good because we know we won't delete anything we need. Bad because it requires an operator noticing and making a configuration change. The cost of them not noticing is we stop removing unneeded things from HDFS, so the cluster could die by exhausting available space.
2) We could use `MOB_MINIMUM_FILE_AGE_TO_ARCHIVE_KEY` as a floor. Say we use an archive threshold of 2x the time to build the list of active mobs or `MOB_MINIMUM_FILE_AGE_TO_ARCHIVE_KEY`, whichever is greater. Good because don't risk cluster death if no one notices. Maybe bad because of how large of a window we end up with? Oh! also bad because if it takes us a long time to get through iterating on the mob directory we might exhaust the min age window still.
3) Instead of using `now` when we start iterating over the mob directory, we use `now` as of before we start iterating over the hfile directories. Good because again no risk of cluster death if no one notices. Also easier to explain/reason about than the variable look-behind of number 2.

what do you think?

Reply:
We can record the time needs to build the set of all active MOB files as T and use 2xT as minimum age to archive.

Reply:
It means we do not need MOB_MINIMUM_FILE_AGE_TO_ARCHIVE_KEY anymore.

Reply:
This is new algorithm:
`// We check only those MOB files, which creation time is less 
      // than maxTimeToArchive. This is a current time - 1h. 1 hour gap 
      // gives us full confidence that all corresponding store files will
      // exist at the time cleaning procedure begins and will be examined.
      // So, if MOB file creation time is greater than this maxTimeToArchive,
      // this will be skipped and won't be archived.
      long maxCreationTimeToArchive = EnvironmentEdgeManager.currentTime() - 3600000;
      LOG.info("Only MOB files whose creation time less than {} will be archived", 
        maxCreationTimeToArchive);`

The idea is to set maxCreationTimeToArchive as now() -1h. now() - when cleaning chore starts. 1h gap gives us  guarantee that all corresponding store files exist at the time we start cleaning. It is hard to imagine, that MOB file and store creation are more that 1h apart.
Then I just check MOB file creation time and if its less  than this threshold - it can be archived.